mteb/models/cde-small-v1_model.py

```diff
@@ -0,0 +1,34 @@
+from __future__ import annotations
+
+import mteb
+
+model = mteb.get_model(
```

Comment:
I'm not sure if I understand this correctly, but it seems like you did not add a model implementation or model metadata for CDEs. I'm also unsure whether this would work or not. I believe their official guide on [how to use CDE](https://huggingface.co/jxm/cde-small-v1) is a bit more complicated than this, since they have a first and a sceond stage in all of their guides where they first produce a corpus embedding and then pass it along to the model when embedding new documents.

Reply:
They have evaluation script, but it a bit complecated https://github.com/jxmorris12/cde/blob/0de4e6c116c8e8223075a2b56277d69e04a2ab7c/evaluate_mteb.py#L26

Reply:
I see, but I guess it's a better choice still not to implement the model incorrectly here, and maybe just add metadata on it, then ask the CDE team to upload their results to the results repository.
I don't see too much value in adding a script here, that does not use CDEs as they are supposed to be used

Reply:
I agree with you. I added it evaluation script just for information and show author's implementation

Reply:
You need to check if your results are matching with results reported by the authors. You can check results in https://huggingface.co/spaces/mteb/leaderboard

Reply:
@YashDThapliyal When you run `get_model_meta`, it can only do so much. It fetches some information about the model, but likely won't have any information on the supported languages, release date, license, etc.
I'd encourage you to add a metadata object for the model and obtain this information from their paper or their HF repo readme.

I'm pretty sure the results are not entirely correct, as their use of the model includes producing a contextual corpus embedding, and only then embedding documents. I think you should take a look at the link @Samoed sent for their eval script.

I might be wrong. If your results match with theirs, then sorry for the trouble. I'd look into this though

Reply:
@Samoed @x-tabdeveloping Thank you for highlighting the corpus document note I will take a look at that and ensure that the model has been implemented correctly—I’ll make sure to update the model metadata with additional information as needed. 

Once the implementation is complete, I’ll compare the results with the existing ones to ensure alignment.

Quick question: What would be the best approach to define the model revision in this context?

Thank you for your guidance!

Best regards,
Yash

Reply:
Thanks! I'd take the last commit's ID on their HF repo as revision.

Reply:
<img width="700" alt="Screenshot 2024-12-23 at 3 46 41 PM" src="https://github.com/user-attachments/assets/fd84c4e4-45b2-4dba-b313-4e09297df95e" />

@x-tabdeveloping @Samoed I tried to make a model card, wanted to show it to you for feedback on if anything seems odd/off. I used the bge_models.py as a reference on how the ModelMeta Card should be displayed.

I wanted to ask now, if I should actually implement CDE (by following the code examples on HF) and then compare the results with the ones published already or should we get in contact with CDE team and ask them to upload their results in the results directory. (wanted to clarify based off of the discussion earlier)

Thank you very much!
Yash Thapliyal

Reply:
Yes, we should implement it. ModelCard looks fine, except zero_shot benchmark, because in main it is already removed.

Reply:
Gotcha, ill work on implementing cde (via the code examples that they have) and ill remove the zero_shot benchmark.

Thanks!

Reply:
Hi @KennethEnevoldsen, @x-tabdeveloping, @Samoed, @isaac-chung,

I’ve added the model implementation, meta card, and results. You can find them here: https://github.com/YashDThapliyal/mteb/tree/main/mteb/models/results/jxm__cde-small-v1

Could you please review these and let me know if they make sense? Additionally, any suggestions for next steps would be appreciated. If everything looks good, should I proceed with a new PR or close this one and start fresh?

Thanks!

Reply:
@YashDThapliyal since @Samoed and @x-tabdeveloping is reviewing this thread I will leave it to them. There is many PRs and sadly we can't all review everything. If the thread has been dead for three days or more feel free to tag me

Reply:
Your results don't seem to match the authors'. I think you should refer to the authors' implementation to see what's missing

Reply:
@Samoed 

Thank you for your feedback! I’d like to clarify and address the observed discrepancies to ensure alignment with expectations.

---

#### **1. Comparison to Authors' Results**
To better understand the differences in results, could you provide more details on how the comparisons were made? Specifically:
- **What metrics were compared?**  
  Were these metrics task-specific scores or overall aggregate scores from the MTEB evaluation?
- **Where are the original authors' results documented?**  
  Is there a specific publication, repository, or script output that contains these results for reference? I tried looking around but didn't come across anything valuable, but maybe im just not looking for the right thing. 

---

#### **2. Analysis of Discrepancies**
Upon reviewing the author's implementation in the evaluation script, here are the key differences I identified that likely contribute to the observed discrepancies:

1. **Model Loading**:
   - **Original Implementation**: Loads a fine-tuned model checkpoint using `analyze_utils`, which may include task-specific adaptations and configurations.
   - **My Implementation**: Directly loads the model (`jxm/cde-small-v1`) via `SentenceTransformer`, missing the fine-tuned settings.

2. **Dataset and Preprocessing**:
   - **Original Implementation**: Dynamically constructs datasets with task-specific sampling, applies prefixes (e.g., `query_prefix` and `document_prefix`), and uses transductive embeddings.
   - **My Implementation**: Uses a simplified random corpus (`random_strings_cde.txt`), which may not be representative or directly comparable.

3. **Prefixes and Embedding Normalization**:
   - **Original Implementation**: Includes task-specific query/document prefixes and normalizes embeddings for clustering tasks.
   - **My Implementation**: Does not fully replicate these configurations, which can influence results.

4. **Batch Size and Device Configurations**:
   - **Original Implementation**: Uses a large batch size (512) and optimized GPU settings (e.g., `torch.autocast`).
   - **My Implementation**: Uses smaller batch sizes (8), which may introduce noise in embeddings. My computer doesn't have a GPU, so I don't think using such a large batch size is smart, I can possibly update the batch to 32 and see if thats feasible.

5. **Evaluation Settings**:
   - **Original Implementation**: Customizes evaluation splits (`dev` or `test`), chunk sizes, and additional configurations for specific tasks.
   - **My Implementation**: Generalizes these settings, potentially contributing to metric differences.

---

I wanted to get your thoughts on which of these differences are worth addressing and feasible to resolve. One idea I had was to simply copy and paste the evaluation script the author used and modify it to work without a GPU. However, at that point, does it make sense for me to implement the script myself, or would it be more practical to ask the author to upload their results directly?

Would love your perspective on the best way to proceed.

Thanks for all of your help and guidance, I really appreciate it!
Yash

Reply:
1. You can find the authors' results in their model card.  
2. Yes, there are some differences in implementation, but I believe the results can match.  

@jxmorris12, could you help integrate your model into MTEB?

Reply:
Hi!  I can help integrate the model into MTEB in the new year. Your implementation doesn't use the task-specific embeddings, for one thing, which is known to decrease performance (see Table 2 of [the paper](https://arxiv.org/pdf/2410.02525)). That said, your results look significantly less performant than even that number, so I suspect something else is wrong.

Reply:
@jxmorris12 Awesome, I look forward to working with you in the new year!